_posts/2018-11-01-http-part6.markdown

```diff
@@ -0,0 +1,514 @@
+
+    fun close() {
+        try {
+            socket.close()
+        } catch (e: Exception) {
+            logger.warn("Could not close the connection", e)
+        }
+    }
+
+    fun isClosed(): Boolean = socket.isClosed
+
+    fun getInputStream(): InputStream = socket.getInputStream()
+            ?.let(::ActivityTrackingInputStream)
+            ?: throw ConnectionException("Could not obtain stream. Is socket closed?")
+
+    fun getOutputStream(): OutputStream = socket.getOutputStream()
+            ?.let(::ActivityTrackingOutputStream)
+            ?: throw ConnectionException("Could not obtain stream. Is socket closed?")
+
+    private inner class ActivityTrackingInputStream(val inputStream: InputStream): InputStream() {
+
+        override fun read(): Int = inputStream.read().also { checks = 0 }
+        override fun read(b: ByteArray?): Int = inputStream.read(b).also { checks = 0 }
+        override fun read(b: ByteArray?, off: Int, len: Int): Int = inputStream.read(b, off, len).also { checks = 0 }
+        override fun skip(n: Long): Long = inputStream.skip(n)
+        override fun available(): Int = inputStream.available()
+        override fun reset() = inputStream.reset()
+        override fun close() = inputStream.close()
+        override fun mark(readlimit: Int) = inputStream.mark(readlimit)
+        override fun markSupported(): Boolean = inputStream.markSupported()
+    }
+
+    private inner class ActivityTrackingOutputStream(val outputStream: OutputStream): OutputStream() {
+
+        override fun write(b: Int) = outputStream.write(b).also { checks = 0 }
+        override fun write(b: ByteArray?) = outputStream.write(b).also { checks = 0 }
+        override fun write(b: ByteArray?, off: Int, len: Int) = outputStream.write(b, off, len).also { checks = 0 }
+        override fun flush() = outputStream.flush()
+        override fun close() = outputStream.close()
+    }
+}
+
+open class ConnectionException(msg: String, throwable: Throwable? = null): RuntimeException(msg, throwable)
+```
+How should we tell whether the connection is no longer active? It would be so easy to save the timestamp of the last activity and compare it to `keepAliveTimeout`, but calling `System.nanoTime()` on every byte send/received would create an overhead. Instead, we create a counter (1) that is incremented on every inactivity check (2).  We know how often this method will be called because of the `checkKeepAliveInternal` argument, so we can compute (3) after how many checks the connection is no longer active.
+If there is any activity, then `ActivityTrackingInputStream` and `ActivityTrackingOutputStream` decorators will reset the counter.
+
+This solution is not ideal, because we have to trust the client of this class to run the `isInactive()` method with the `checkKeepAliveInterval` interval. On the other hand, the `Connection` class won’t be visible to the end user of our `SocketClient` so I think we can live with that.
```

Comment:
solution is not ideal -> not perferct

Reply:
What's the difference between ideal and perfect in this context?

I think that "ideal" means something that would meet Kuba's expectations. Perfect means without flaws. So I think that it's up to the author to decide. 

I think that the solution is definitely not perfect, but it might be ideal for someone.

Reply:
I looked up the definitions.

perfect "having all the required or desirable elements, qualities, or characteristics; as good as it is possible to be."
ideal "satisfying one's conception of what is perfect; most suitable."

I think you're right Krzysiek. I thought that perfect = ideal and just preferred the sound of "ideal" :)